libraries/Microsoft.Bot.Connector/Authentication/MsalAppCredentials.cs

```diff
@@ -0,0 +1,227 @@
+#pragma warning disable VSTHRD103 // Call async methods when in an async method
+                acquired = tokenRefreshSemaphore.Wait(SemaphoreTimeout);
+#pragma warning restore VSTHRD103 // Call async methods when in an async method
+
+                // If we are allowed to enter the semaphore, acquire the token.
+                if (acquired)
+                {
+                    // Note that in MSAL, we dont pass resources anymore, and we instead pass scopes. To be recognized by MSAL, we append the '/.default' to the scope.
+                    // Scope requirements described in MSAL migration spec: https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-migration.
+                    const string scopePostFix = "/.default";
+                    var scope = _scope ?? OAuthScope;
+
+                    if (!scope.EndsWith(scopePostFix, StringComparison.OrdinalIgnoreCase))
+                    {
+                        scope = $"{scope}{scopePostFix}";
+                    }
+
+                    // Acquire token async using MSAL.NET
+                    // This will use the cache from the application cache of the MSAL library, no external caching is needed.
+                    var msalResult = await _clientApplication
+                        .AcquireTokenForClient(new[] { scope })
+                        .WithAuthority(_authority ?? OAuthEndpoint, _validateAuthority)
+                        .WithForceRefresh(forceRefresh)
+                        .ExecuteAsync().ConfigureAwait(false);
+
+                    // This means we acquired a valid token successfully. We can make our retry policy null.
+                    return new AuthenticatorResult()
+                    { 
+                        AccessToken = msalResult.AccessToken,
+                        ExpiresOn = msalResult.ExpiresOn
+                    };
+                }
+                else
+                {
+                    // If the token is taken, it means that one thread is trying to acquire a token from the server.
+                    // Throttle this request to allow the currently running request to fulfill and then let this one get the result from the cache.
+                    throw new ThrottleException() { RetryParams = RetryParams.DefaultBackOff(0) };
+                }
+            }
+            finally
+            {
+                // Always release the semaphore if we acquired it.
+                if (acquired)
+                {
+                    ReleaseSemaphore();
+                }
+            }
+        }
+
+        private void ReleaseSemaphore()
+        {
+            try
+            {
+                tokenRefreshSemaphore.Release();
+            }
+            catch (SemaphoreFullException)
+            {
+                // We should not be hitting this after switching to SemaphoreSlim, but if we do hit it everything will keep working.
```

Comment:
the comment is confusing... you _are_ using SemaphoreSlim???

Reply:
It's outdated, I'll remove. thanks!

Reply:
done, I changed the comment.